src/Lykke.Service.ClientAccountRecovery.Services/ChallengeManager.cs

```diff
@@ -0,0 +1,79 @@
+﻿using System;
+using System.Threading.Tasks;
+using JetBrains.Annotations;
+using Lykke.Service.ClientAccountRecovery.Core.Domain;
+using Lykke.Service.ClientAccountRecovery.Core.Services;
+using Action = Lykke.Service.ClientAccountRecovery.Core.Domain.Action;
+
+namespace Lykke.Service.ClientAccountRecovery.Services
+{
+    [UsedImplicitly]
+    public class ChallengeManager : IChallengeManager
+    {
+        private readonly IChallengeValidatorFactory _challengeValidatorFactory;
+        private readonly ISelfieNotificationSender _selfieNotificationSender;
+
+        public ChallengeManager(IChallengeValidatorFactory challengeValidatorFactory, ISelfieNotificationSender selfieNotificationSender)
+        {
+            _challengeValidatorFactory = challengeValidatorFactory;
+            _selfieNotificationSender = selfieNotificationSender;
+        }
+
+        public async Task<bool> ExecuteAction(Challenge challenge, Action action, string code, IRecoveryFlowService flow)
+        {
+            var switcher = (challenge, action);
+            switch (switcher)
```

Comment:
ugly construction, not readable at all

Probably better to extract it as some "rules" and have factory.

Reply:
No obvious benefits for this particular case.